Add VSISync() to synchronize source and target files/directories

rouault opened the pull request.

rcoup commented:
1. I think this should compare file sizes before calculating md5sums, the latter are relatively expensive to calculate on large files. And if the size differs the hashes definitely will.

2. S3 ETags for files uploaded via multipart upload are [not straight md5sums](https://www.savjee.be/2015/10/Verifying-Amazon-S3-multi-part-uploads-with-ETag-hash/).
Calculating it isn't complex, but you need to know the chunk size the multipart upload used to get the matching local hash, and that's not available from S3. The AWS CLI uses an 8MB chunk size by default (but it's customisable). Apparently you can [force a recalculation of the ETag by changing metadata attributes](https://stackoverflow.com/a/19033497/2662), but only works for files <5GB.

3. ETags for S3 with KMS server side encryption aren't MD5s of the content either. 

For 2 & 3, I'm not sure there's much we can do, except possibly support an additional metadata attribute (`Content-MD5`?) containing a straight MD5.

rcoup commented:
Looks like `aws s3 sync` uses size + last modified time and doesn't touch etags: 
https://github.com/aws/aws-cli/blob/develop/awscli/customizations/s3/syncstrategy/base.py#L226-L256

rouault commented:
@rcoup Thanks for your feedback. I've modified the default SYNC_STRATEGY to be timestamp based applying the 'aws s3 sync' rules (which aren't totally bullet proof but quite reasonable). SYNC_STRATEGY=ETAG can be passed as an option with the limitations you underlined and I've added in the doc)
